src/kontrol/kdist/cheatcodes.md

```diff
@@ -907,6 +913,53 @@ Otherwise, throw an error for any other call to the Foundry contract.
       [owise]
 ```
 
+```k
+    rule [foundry.call.mockCall]:
+         <k> #call_foundry SELECTOR ARGS
+          => #loadAccount #asWord(#range(ARGS, 0, 32)) 
+          ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, #asWord(#range(ARGS, 32, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 32, 32)), 32))) #range(ARGS, #asWord(#range(ARGS, 64, 32)) +Int 32, #asWord(#range(ARGS, #asWord(#range(ARGS, 64, 32)), 32)))
+         ...
+         </k> 
+      requires SELECTOR ==Int selector ( "mockCall(address,bytes,bytes)" )
+
+    //rule [foundry.call.mockCallWithMsgValue]:
+    //     <k> #call_foundry SELECTOR ARGS
+    //        => #loadAccount #asWord(#range(ARGS, 0, 32))
+    //        ~> #setMockCall #asWord(#range(ARGS, 0, 32)) #range(ARGS, 96, #asWord(#range(ARGS, 64, 32))) false ... </k>
+    //  requires SELECTOR ==Int selector ( "mockCall(address,uint256,bytes,bytes)" )
+```
+
+```k
+    rule [foundry.set.mockCall]:
+         <k> #next [ _OP:CallOp ] ~> (. =>  #setLocalMem RETSTART RETWIDTH RETURNDATA) ~> #execute ... </k>
+         <localMem> LM </localMem>
+         <wordStack> _ : ACCTTO : _ : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
+         <mockCall>
+           <mockAddress> ACCTTO </mockAddress>
+           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
+         </mockCall>
+         requires #range(LM, ARGSTART, ARGWIDTH) ==K #range(CALLDATA,0,ARGWIDTH)
+      [priority(30)]
```

Comment:
I think that here you could have something like this instead. Basically saying that if you have a mocked value for a given address, skip executing the `CALL` opcode and just set the mocked value in the local memory and output cell.
```suggestion
    rule [foundry.set.mockCall]:
         <k> #next [ _OP:CallOp ] =>  #setLocalMem RETSTART RETWIDTH RETURNDATA  ... </k>
         <localMem> LM </localMem>
         <wordStack> _ : ACCTTO : _ : ARGSTART : ARGWIDTH : RETSTART : RETWIDTH : _WS </wordStack>
         <mockCall>
           <mockAddress> ACCTTO </mockAddress>
           <mockValues>...  CALLDATA |-> RETURNDATA ...</mockValues>
         </mockCall>
         requires #range(LM, ARGSTART, ARGWIDTH) ==K #range(CALLDATA,0,ARGWIDTH)
      [priority(30)]
```

Reply:
That was my first approach, but then I noticed (and tested) that actually if the function called changes the state of the contract, those changes are made, which means that the function is executed, but instead, it returns the value provided in the `mockCall`.

Reply:
I see. Then you can leave them as they are. :+1: